### PR TITLE
Refine the implementation of TLS

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

1. tls: Merge tls_xxx into pthread_keyxxx 
2. libc/tls: Make tls_get_info as the pulibc function instead up_tls_info 
3. sched/tls: Add tls_init_info and tls_dup_info 
4. sched/tls: Add task_init_info and task_uninit_info 

## Impact
Code refactor

## Testing
ostest

### CODE REVIEW

File: libs/libc/tls/Make.defs
Comment (pkarashchenko) [inline libs/libc/tls/Make.defs:24]: Maybe can move `tls_destruct.c` and name it `pthread_keydestruct.c`?

**Comment by pkarashchenko** [comment]
Please rebase your changes

File: libs/libc/tls/Make.defs
Comment (xiaoxiang781216) [reply libs/libc/tls/Make.defs:24]: But functions inside tls_destruct.c isn't specified by POSIX.

File: libs/libc/pthread/pthread_keycreate.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keycreate.c:80]:
```suggestion
  DEBUGASSERT(info != NULL);
```

File: libs/libc/pthread/pthread_keydelete.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keydelete.c:71]:
```suggestion
      mask = (tls_ndxset_t)1 << key;
```

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keydelete.c]: Done.

File: libs/libc/pthread/pthread_keycreate.c
Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c]: Done.

**Review by pkarashchenko — APPROVED** [review]

File: libs/libc/pthread/pthread_setspecific.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_setspecific.c:80]: What is the point of `DEBUGASSERT` + `if` the same condition?
Also do we need to check `key >= 0`?

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_setspecific.c:80]: Compiler may complain the check(key >= 0) is redundant since key is an unsigned int.

Comment (pkarashchenko) [reply libs/libc/pthread/pthread_setspecific.c:80]: But I see
```
#ifndef __PTHREAD_KEY_T_DEFINED
typedef int pthread_key_t;
#define __PTHREAD_KEY_T_DEFINED 1
#endif
```
Are you using custom `pthread_key_t` definition as unsigned?

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_setspecific.c]: Ok, my editor direct me to the wrong definition(glibc version). All code check `key >= 0` now.

**Comment by xiaoxiang781216** [comment]
depend on: https://github.com/apache/incubator-nuttx-apps/pull/1173

**Comment by davids5** [comment]
@xiaoxiang781216  are the comments https://github.com/apache/incubator-nuttx/blob/a10c4dde56134f54974c4bf7f57c49e3cd1eeadb/libs/libc/tls/Kconfig#L19  still valid?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 are the comments
> 
> https://github.com/apache/incubator-nuttx/blob/a10c4dde56134f54974c4bf7f57c49e3cd1eeadb/libs/libc/tls/Kconfig#L19
> 
> still valid?

Yes, it's here:
https://github.com/apache/incubator-nuttx/pull/6337/files#diff-2c833ffaa18374fab0d57499dc593937167b198a630cb4e9068f654aaebdc42dR308
I just mov the same code from include/nuttx/arch.h to include/nuttx/tls.h

File: libs/libc/pthread/pthread_keycreate.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keycreate.c:108]: I think `ret = OK;` is missing before `break;` ;)

File: libs/libc/pthread/pthread_keycreate.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keycreate.c:92]: I think we need to wait for https://github.com/apache/incubator-nuttx/pull/6320 to be merged first as according to POSIX: `The pthread_key_create() function shall not return an error code of [EINTR]`

File: libs/libc/pthread/pthread_keydelete.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keydelete.c:73]: I think we need to wait for https://github.com/apache/incubator-nuttx/pull/6320 to be merged first as according to POSIX: `The pthread_key_delete() function shall not return an error code of [EINTR].`

File: libs/libc/pthread/pthread_keycreate.c
Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:92]: Do you want to replace sem with nxmutex? it's better to wait @anjiahao1 do the replacement in the batch.

File: libs/libc/pthread/pthread_keydelete.c
Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keydelete.c:73]: @pkarashchenko could you merge https://github.com/apache/incubator-nuttx-apps/pull/1173 first to unblock the ci?

Comment (pkarashchenko) [reply libs/libc/pthread/pthread_keydelete.c:73]: Done

File: libs/libc/pthread/pthread_keycreate.c
Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:109]: Done.

File: libs/libc/pthread/pthread_keycreate.c
Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:92]: @anjiahao1 is preparing the patch to replace sem lock with mutex, so let's wait his patch instead.

File: libs/libc/pthread/pthread_keycreate.c
Comment (pkarashchenko) [inline libs/libc/pthread/pthread_keycreate.c:86]: @xiaoxiang781216 could you please switch to nxmutex or add EINTR handling?

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:86]: I prefer to wait the patch which replace all sem lock with mutex lock instead, also it's the same code before merge from tls_alloc.c to pthread_keycreate.c

Comment (pkarashchenko) [reply libs/libc/pthread/pthread_keycreate.c:86]: Please address the `EINTR` handling that should not be returned by `pthread_key_create`

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:86]: Three reasons I want to keep the current change:

1. The old code doesn't handle EINTR too
2. Many other place use semaphore as lock without checking EINTR
3. @anjiahao1 is preparing the patch to convert all item 2 to mutex

So, I prefer we fix the problem in a unified(not adhoc) approach.

Comment (pkarashchenko) [reply libs/libc/pthread/pthread_keycreate.c:86]: 1. Old code also has a bug that violates POSIX.
2. Other places are most probably not the POSIX API use cases that prohibit EINTR from being returned.
3. So do you propose to wait with this change until the mutex change is introduced?

Comment (xiaoxiang781216) [reply libs/libc/pthread/pthread_keycreate.c:86]: > 1. Old code also has a bug that violates POSIX.

Yes, but it should be handled by new patch. It's better that one patch just fix one thing.

> 2. Other places are most probably not the POSIX API use cases that prohibit EINTR from being returned.

It's not good to return EINTR in case of locking(not event or signal) whether is required by POSIX or not.

> 3. So do you propose to wait with this change until the mutex change is introduced?

No, I mean we can merge this pr first and:

1. Wait the mutex change to fix all places in a unify approach or
2. Create a new patch to just fix pthread_key_create/pthread_key_delete

Comment (pkarashchenko) [reply libs/libc/pthread/pthread_keycreate.c:86]: Ok

**Review by pkarashchenko — APPROVED** [review]

**Comment by masayuki2009** [comment]
@xiaoxiang781216 

I noticed that rv-virt:knsh64 failed.

```
$ qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt -bios none -kernel ./nuttx/nuttx
ABC[    0.000000] up_assert: Assertion failed at file:init/nx_bringup.c line: 297 task: Idle Task
[    0.000000] up_assert: Assertion failed at file:common/riscv_swint.c line: 189 task: Idle Task
[    0.000000] riscv_registerdump: EPC: 00000000800071de
[    0.000000] riscv_registerdump: A0: 0000000000000000 A1: 0000000000000000 A2: 0000000010000000 A3: 0000000000000001
[    0.000000] riscv_registerdump: A4: 0000000000000001 A5: 0000000000000000 A6: 0000000000000074 A7: 000000000000006a
[    0.000000] riscv_registerdump: T0: 00000000000001ff T1: 000000000000006c T2: 000000000000002d T3: 0000000000000068
[    0.000000] riscv_registerdump: T4: 0000000000000009 T5: 000000000000002a T6: 000000000000002e
[    0.000000] riscv_registerdump: S0: 0000000080200928 S1: 0000000080206b30 S2: 0000000080200928 S3: 0000000080200918
[    0.000000] riscv_registerdump: S4: 0000000080201050 S5: 0000000000000000 S6: 0000000000000000 S7: 0000000000000000
[    0.000000] riscv_registerdump: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[    0.000000] riscv_registerdump: SP: 0000000080206b20 FP: 0000000080200928 TP: 0000000000000000 RA: 00000000800071de
[    0.000000] riscv_dumpstate: sp:     0000000080200850
[    0.000000] riscv_dumpstate: IRQ stack:
[    0.000000] riscv_dumpstate:   base: 00000000802000f0
[    0.000000] riscv_dumpstate:   size: 0000000000000800
[    0.000000] riscv_stackdump: 0000000080200840: 802000f0 00000000 80006cbc 00000000 000000bd 00000000 80015138 00000000
[    0.000000] riscv_stackdump: 0000000080200860: deadbeef deadbeef 00000000 00000000 80201050 00000000 80200918 00000000
[    0.000000] riscv_stackdump: 0000000080200880: 80201050 00000000 80206a18 00000000 80206a18 00000000 800043ac 00000000
[    0.000000] riscv_stackdump: 00000000802008a0: deadbeef deadbeef 800015f6 00000000 80200928 00000000 80201050 00000000
[    0.000000] riscv_stackdump: 00000000802008c0: 80206a18 00000000 80001812 00000000 deadbeef deadbeef 00000100 00000000
[    0.000000] riscv_dumpstate: sp:     0000000080206b20
[    0.000000] riscv_dumpstate: User stack:
[    0.000000] riscv_dumpstate:   base: 0000000080206030
[    0.000000] riscv_dumpstate:   size: 0000000000000bd0
[    0.000000] riscv_dumpstate: User Stack
[    0.000000] riscv_stackdump: 0000000080206b20: 00000000 00000000 80006e50 00000000 00000129 00000000 80015280 00000000
[    0.000000] riscv_stackdump: 0000000080206b40: 00000000 00000000 00000000 00000000 00000000 00000000 80200918 00000000
[    0.000000] riscv_stackdump: 0000000080206b60: 80201064 00000000 80201068 00000000 80200928 00000000 800043ac 00000000
[    0.000000] riscv_stackdump: 0000000080206b80: 80200928 00000000 80001b48 00000000 80200928 00000000 80016400 00000000
[    0.000000] riscv_stackdump: 0000000080206ba0: 80200928 00000000 80001a94 00000000 80400000 00000000 00400000 00000000
[    0.000000] riscv_stackdump: 0000000080206bc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.000000] riscv_stackdump: 0000000080206be0: 00000000 00000000 80000d72 00000000 00000000 00000000 80000046 00000000
[    0.000000] riscv_showtasks:    PID    PRI      USED     STACK   FILLED    COMMAND
[    0.000000] riscv_showtasks:   ----   ----       744      2048    36.3%    irq
[    0.000000] riscv_dump_task:      0      0      1184      3024    39.1%    Idle Task

```

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> 
> I noticed that rv-virt:knsh64 failed.
> 
> ```
> $ qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt -bios none -kernel ./nuttx/nuttx
> ABC[    0.000000] up_assert: Assertion failed at file:init/nx_bringup.c line: 297 task: Idle Task
> [    0.000000] up_assert: Assertion failed at file:common/riscv_swint.c line: 189 task: Idle Task
> [    0.000000] riscv_registerdump: EPC: 00000000800071de
> [    0.000000] riscv_registerdump: A0: 0000000000000000 A1: 0000000000000000 A2: 0000000010000000 A3: 0000000000000001
> [    0.000000] riscv_registerdump: A4: 0000000000000001 A5: 0000000000000000 A6: 0000000000000074 A7: 000000000000006a
> [    0.000000] riscv_registerdump: T0: 00000000000001ff T1: 000000000000006c T2: 000000000000002d T3: 0000000000000068
> [    0.000000] riscv_registerdump: T4: 0000000000000009 T5: 000000000000002a T6: 000000000000002e
> [    0.000000] riscv_registerdump: S0: 0000000080200928 S1: 0000000080206b30 S2: 0000000080200928 S3: 0000000080200918
> [    0.000000] riscv_registerdump: S4: 0000000080201050 S5: 0000000000000000 S6: 0000000000000000 S7: 0000000000000000
> [    0.000000] riscv_registerdump: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
> [    0.000000] riscv_registerdump: SP: 0000000080206b20 FP: 0000000080200928 TP: 0000000000000000 RA: 00000000800071de
> [    0.000000] riscv_dumpstate: sp:     0000000080200850
> [    0.000000] riscv_dumpstate: IRQ stack:
> [    0.000000] riscv_dumpstate:   base: 00000000802000f0
> [    0.000000] riscv_dumpstate:   size: 0000000000000800
> [    0.000000] riscv_stackdump: 0000000080200840: 802000f0 00000000 80006cbc 00000000 000000bd 00000000 80015138 00000000
> [    0.000000] riscv_stackdump: 0000000080200860: deadbeef deadbeef 00000000 00000000 80201050 00000000 80200918 00000000
> [    0.000000] riscv_stackdump: 0000000080200880: 80201050 00000000 80206a18 00000000 80206a18 00000000 800043ac 00000000
> [    0.000000] riscv_stackdump: 00000000802008a0: deadbeef deadbeef 800015f6 00000000 80200928 00000000 80201050 00000000
> [    0.000000] riscv_stackdump: 00000000802008c0: 80206a18 00000000 80001812 00000000 deadbeef deadbeef 00000100 00000000
> [    0.000000] riscv_dumpstate: sp:     0000000080206b20
> [    0.000000] riscv_dumpstate: User stack:
> [    0.000000] riscv_dumpstate:   base: 0000000080206030
> [    0.000000] riscv_dumpstate:   size: 0000000000000bd0
> [    0.000000] riscv_dumpstate: User Stack
> [    0.000000] riscv_stackdump: 0000000080206b20: 00000000 00000000 80006e50 00000000 00000129 00000000 80015280 00000000
> [    0.000000] riscv_stackdump: 0000000080206b40: 00000000 00000000 00000000 00000000 00000000 00000000 80200918 00000000
> [    0.000000] riscv_stackdump: 0000000080206b60: 80201064 00000000 80201068 00000000 80200928 00000000 800043ac 00000000
> [    0.000000] riscv_stackdump: 0000000080206b80: 80200928 00000000 80001b48 00000000 80200928 00000000 80016400 00000000
> [    0.000000] riscv_stackdump: 0000000080206ba0: 80200928 00000000 80001a94 00000000 80400000 00000000 00400000 00000000
> [    0.000000] riscv_stackdump: 0000000080206bc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
> [    0.000000] riscv_stackdump: 0000000080206be0: 00000000 00000000 80000d72 00000000 00000000 00000000 80000046 00000000
> [    0.000000] riscv_showtasks:    PID    PRI      USED     STACK   FILLED    COMMAND
> [    0.000000] riscv_showtasks:   ----   ----       744      2048    36.3%    irq
> [    0.000000] riscv_dump_task:      0      0      1184      3024    39.1%    Idle Task
> ```

let me try it.

**Comment by xiaoxiang781216** [comment]
@masayuki2009 the assertion happen here:
```
  ret = exec_spawn(CONFIG_INIT_FILEPATH, argv, NULL,
                   CONFIG_INIT_SYMTAB, CONFIG_INIT_NEXPORTS, &attr);
  DEBUGASSERT(ret >= 0);
```
But I can't find init in my apps/bin/ folder. Could you point me the setup document of knsh64?

**Comment by masayuki2009** [comment]
@xiaoxiang781216 

>But I can't find init in my apps/bin/ folder. Could you point me the setup document of knsh64?

The crash messages were incorrect.

```
$ qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt -bios none -kernel ./nuttx
ABC
NuttShell (NSH) NuttX-10.3.0
nsh> hello
Hello, World!!
[    0.000000] riscv_exception: EXCEPTION: Store/AMO access fault. MCAUSE: 0000000000000007
[    0.000000] riscv_exception: PANIC!!! Exception = 0000000000000007
[    0.000000] up_assert: Assertion failed at file:common/riscv_exception.c line: 89 task: hello
[    0.000000] riscv_registerdump: EPC: 0000000080005c3c
[    0.000000] riscv_registerdump: A0: 0000000000000001 A1: 0000000080208038 A2: 0000000000000019 A3: 0000000080206c38
[    0.000000] riscv_registerdump: A4: 0000000080208018 A5: 0000000080208058 A6: 0000000000000101 A7: 0000000000000000
[    0.000000] riscv_registerdump: T0: 000000008000399c T1: 0000000000000000 T2: 0000000000000000 T3: 0000000000000000
[    0.000000] riscv_registerdump: T4: 0000000000000000 T5: 0000000000000000 T6: 0000000000000000
[    0.000000] riscv_registerdump: S0: 0000000080208040 S1: 0000000080206c00 S2: ffffffffffffffe0 S3: fffffffffffffff8
[    0.000000] riscv_registerdump: S4: ffffffffffffffe0 S5: 0000000000000000 S6: 0000000000000000 S7: 0000000000000000
[    0.000000] riscv_registerdump: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[    0.000000] riscv_registerdump: SP: 0000000080209d90 FP: 0000000080208040 TP: 0000000000000000 RA: 0000000080005b60
[    0.000000] riscv_dumpstate: sp:     0000000080200830
[    0.000000] riscv_dumpstate: IRQ stack:
[    0.000000] riscv_dumpstate:   base: 00000000802000f0
[    0.000000] riscv_dumpstate:   size: 0000000000000800
[    0.000000] riscv_stackdump: 0000000080200820: 802000f0 00000000 80006cbc 00000000 00000059 00000000 80014ee8 00000000
[    0.000000] riscv_stackdump: 0000000080200840: deadbeef deadbeef 00000000 00000000 ffffffe0 ffffffff fffffff8 ffffffff
[    0.000000] riscv_stackdump: 0000000080200860: ffffffe0 ffffffff 80209c88 00000000 00000007 00000000 800043ac 00000000
[    0.000000] riscv_stackdump: 0000000080200880: 00000009 00000000 80001464 00000000 00000000 00000000 80201050 00000000
[    0.000000] riscv_stackdump: 00000000802008a0: 80209c88 00000000 80001bca 00000000 c0003200 00000000 80000f98 00000000
[    0.000000] riscv_stackdump: 00000000802008c0: 80209d58 00000000 80206c00 00000000 00000007 00000000 80000dfe 00000000
[    0.000000] riscv_dumpstate: sp:     0000000080209d90
[    0.000000] riscv_dumpstate: User stack:
[    0.000000] riscv_dumpstate:   base: 00000000c0004050
[    0.000000] riscv_dumpstate:   size: 00000000000007b0
[    0.000000] riscv_dumpstate: Kernel stack:
[    0.000000] riscv_dumpstate:   base: 0000000080209260
[    0.000000] riscv_dumpstate:   size: 0000000000000c00
[    0.000000] riscv_dumpstate: Kernel Stack
[    0.000000] riscv_stackdump: 0000000080209d80: 00000000 00000000 00040120 00000000 fffffff8 ffffffff 80209120 00000000
[    0.000000] riscv_stackdump: 0000000080209da0: fffffff8 ffffffff 80007a52 00000000 00000000 00000000 00000000 00000000
[    0.000000] riscv_stackdump: 0000000080209dc0: 802090c2 00000000 80208f00 00000000 802090a0 00000000 800127e4 00000000
[    0.000000] riscv_stackdump: 0000000080209de0: 80207ec0 00000000 80010c8e 00000000 802090c2 00000000 802090a0 00000000
[    0.000000] riscv_stackdump: 0000000080209e00: 80208f00 00000000 80011982 00000000 00000000 00000000 00000000 00000000
[    0.000000] riscv_stackdump: 0000000080209e20: 00000000 00000000 80011c54 00000000 00000000 00000000 800039a8 00000000
[    0.000000] riscv_stackdump: 0000000080209e40: 00000000 00000000 800015bc 00000000 c00014de 00000000 00040020 00000000
[    0.000000] riscv_showtasks:    PID    PRI      USED     STACK   FILLED    COMMAND
[    0.000000] riscv_showtasks:   ----   ----       776      2048    37.8%    irq
[    0.000000] riscv_dump_task:      0      0      1328      3024    43.9%    Idle Task
[    0.000000] riscv_dump_task:      1    100      1952      1952   100.0%!   /system/bin/init
[    0.000000] riscv_dump_task:      2    100       536      1968    27.2%    hello

```

It seems that the init task consumes more stack than before.
Actually, it consumed less memory before merging this PR.

```
qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt -bios none -kernel ./nuttx
ABC
NuttShell (NSH) NuttX-10.3.0
nsh> hello
Hello, World!!
nsh> ps
  PID GROUP PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED COMMAND
    0     0   0 FIFO     Kthread N-- Ready              00000000 003024 001328  43.9%  Idle Task
    1     1 100 RR       Task    --- Running            00000000 002000 001736  86.8%! /system/bin/init
nsh> 

```

**Comment by xiaoxiang781216** [comment]
@masayuki2009 it is made by https://github.com/apache/incubator-nuttx/pull/6337/commits/48a0a2ebf3c927c7374b226ac45394df4b877dbe. Could you try #6365?

**Pull request merged**
